File: api/src/main/java/org/openmrs/ProgramWorkflow.java
/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

import org.hibernate.envers.Audited;
import org.openmrs.util.NaturalStrings;

/**
 * ProgramWorkflow
 */
@Audited
public class ProgramWorkflow extends BaseChangeableOpenmrsMetadata {
    
    private static final long serialVersionUID = 1L;

    // ******************
    // Properties
    // ******************
    
    private Integer programWorkflowId;
    private Program program;
    private Concept concept;
    private Set<ProgramWorkflowState> states = new HashSet<>();

    // ******************
    // Constructors
    // ******************
    
    /** Default Constructor */
    public ProgramWorkflow() {}

    /** Constructor with id */
    public ProgramWorkflow(Integer programWorkflowId) {
        setProgramWorkflowId(programWorkflowId);
    }

    // ******************
    // Instance methods
    // ******************

    /**
     * Adds a new {@link ProgramWorkflowState} to this ProgramWorkflow
     * 
     * @param state the {@link ProgramWorkflowState} to add
     */
    public void addState(ProgramWorkflowState state) {
        state.setProgramWorkflow(this);
        states.add(state);
    }

    /**
     * Removes a {@link ProgramWorkflowState} from this ProgramWorkflow
     * 
     * @param state the {@link ProgramWorkflowState} to remove
     */
    public void removeState(ProgramWorkflowState state) {
        if (states.contains(state)) {
            states.remove(state);
            state.setProgramWorkflow(null);
        }
    }

    /**
     * Retires a {@link ProgramWorkflowState}
     * 
     * @param state the {@link ProgramWorkflowState} to retire
     */
    public void retireState(ProgramWorkflowState state) {
        state.setRetired(true);
    }

    /**
     * Returns a {@link ProgramWorkflowState} by primary key id
     * 
     * @param programWorkflowStateId the primary key {@link Integer} id
     * @return the {@link ProgramWorkflowState} or null if not found
     */
    public ProgramWorkflowState getState(Integer programWorkflowStateId) {
        return states.stream()
                     .filter(s -> s.getProgramWorkflowStateId().equals(programWorkflowStateId))
                     .findFirst()
                     .orElse(null);
    }

    /**
     * Returns a {@link ProgramWorkflowState} by Concept
     * 
     * @param concept the Concept to match
     * @return the {@link ProgramWorkflowState} or null if not found
     */
    public ProgramWorkflowState getState(Concept concept) {
        return states.stream()
                     .filter(s -> s.getConcept().equals(concept))
                     .findFirst()
                     .orElse(null);
    }

    /**
     * Returns a {@link ProgramWorkflowState} by Concept name in any {@link Locale}
     * 
     * @param name the Concept name to match
     * @return the {@link ProgramWorkflowState} or null if not found
     */
    public ProgramWorkflowState getState(String name) {
        return states.stream()
                     .filter(s -> s.getConcept().isNamed(name))
                     .findFirst()
                     .orElse(null);
    }

    /**
     * Returns a Set&lt;{@link ProgramWorkflowState}&gt; including all non-retired ProgramWorkflowStates
     * and all retired if <code>includeRetired</code> is true
     * 
     * @param includeRetired if false, returns only non-retired {@link ProgramWorkflowState}
     * @return Set&lt;ProgramWorkflowState&gt; - matching states
     */
    public Set<ProgramWorkflowState> getStates(boolean includeRetired) {
        Set<ProgramWorkflowState> result = new HashSet<>();
        for (ProgramWorkflowState s : states) {
            if (includeRetired || !s.getRetired()) {
                result.add(s);
            }
        }
        return result;
    }

    /**
     * Returns a Set&lt;{@link ProgramWorkflowState}&gt; sorted by {@link ConceptName}
     * 
     * @return Set&lt;ProgramWorkflowState&gt; - sorted states
     */
    public Set<ProgramWorkflowState> getSortedStates() {
        Comparator<String> naturalComparator = NaturalStrings.getNaturalComparator();
        
        Comparator<ProgramWorkflowState> stateComparator = Comparator.comparing(
            s -> s.getConcept().getName().getName(),
            naturalComparator
        );

        Set<ProgramWorkflowState> sorted = new TreeSet<>(stateComparator);
        sorted.addAll(states);
        return sorted;
    }

    /**
     * Returns a List&lt;{@link ProgramWorkflowState}&gt; of possible next states for the given {@link PatientProgram}
     * 
     * @param patientProgram the PatientProgram to check
     * @return List&lt;ProgramWorkflowState&gt; - possible next states
     */
    public List<ProgramWorkflowState> getPossibleNextStates(PatientProgram patientProgram) {
        List<ProgramWorkflowState> result = new ArrayList<>();
        PatientState currentState = patientProgram.getCurrentState(this);
        
        for (ProgramWorkflowState state : getSortedStates()) {
            if (isLegalTransition(currentState == null ? null : currentState.getState(), state)) {
                result.add(state);
            }
        }
        return result;
    }

    /**
     * Checks if transitioning from <code>fromState</code> to <code>toState</code> is allowable.
     * 
     * @param fromState the state to transition from
     * @param toState the state to transition to
     * @return true if the transition is allowable
     */
    public boolean isLegalTransition(ProgramWorkflowState fromState, ProgramWorkflowState toState) {
        // If there's no current state, move into an initial state
        if (fromState == null) {
            return toState.getInitial();
        }

        // Prevent moving into the same state
        return !fromState.equals(toState);
    }

    /** @see Object#toString() */
    @Override
    public String toString() {
        return "ProgramWorkflow(id=" + getProgramWorkflowId() + ")";
    }

    // ******************
    // Property Access
    // ******************

    public Set<ProgramWorkflowState> getStates() {
        return states;
    }

    public void setStates(Set<ProgramWorkflowState> states) {
        this.states = states;
    }

    public Concept getConcept() {
        return concept;
    }

    public void setConcept(Concept concept) {
        this.concept = concept;
    }

    public Program getProgram() {
        return program;
    }

    public void setProgram(Program program) {
        this.program = program;
    }

    public Integer getProgramWorkflowId() {
        return programWorkflowId;
    }

    public void setProgramWorkflowId(Integer programWorkflowId) {
        this.programWorkflowId = programWorkflowId;
    }

    /**
     * @since 1.5
     * @see org.openmrs.OpenmrsObject#getId()
     */
    @Override
    public Integer getId() {
        return getProgramWorkflowId();
    }

    /**
     * @since 1.5
     * @see org.openmrs.OpenmrsObject#setId(java.lang.Integer)
     */
    @Override
    public void setId(Integer id) {
        setProgramWorkflowId(id);
    }

    /**
     * Gets the number of non-retired states
     * 
     * @return the total number of non-retired states
     * @since 1.9
     */
    public int getNonRetiredStateCount() {
        return getStates(false).size();
    }
}

File: api/src/main/kotlin/org/openmrs/ProgramWorkflow.kt
/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs

import org.hibernate.envers.Audited
import org.openmrs.util.NaturalStrings

/**
 * ProgramWorkflow
 */
@Audited
open class ProgramWorkflow : BaseChangeableOpenmrsMetadata {
    
    companion object {
        private const val serialVersionUID = 1L
    }
    
    // ******************
    // Properties
    // ******************
    
    open var programWorkflowId: Int? = null
    
    open var program: Program? = null
    
    open var concept: Concept? = null
    
    open var states: MutableSet<ProgramWorkflowState> = mutableSetOf()
    
    // ******************
    // Constructors
    // ******************
    
    /** Default Constructor */
    constructor()
    
    /** Constructor with id */
    constructor(programWorkflowId: Int?) {
        this.programWorkflowId = programWorkflowId
    }
    
    // ******************
    // Instance methods
    // ******************
    
    /**
     * Adds a new [ProgramWorkflowState] to this ProgramWorkflow
     * 
     * @param state - the [ProgramWorkflowState] to add
     */
    open fun addState(state: ProgramWorkflowState) {
        state.programWorkflow = this
        states.add(state)
    }
    
    /**
     * Removes a [ProgramWorkflowState] from this ProgramWorkflow
     * 
     * @param state - the [ProgramWorkflowState] to remove
     */
    open fun removeState(state: ProgramWorkflowState) {
        if (states.contains(state)) {
            states.remove(state)
            state.programWorkflow = null
        }
    }
    
    /**
     * Retires a [ProgramWorkflowState]
     * 
     * @param state - the [ProgramWorkflowState] to retire
     */
    open fun retireState(state: ProgramWorkflowState) {
        state.retired = true
    }
    
    /**
     * Returns a [ProgramWorkflowState] whose primary key id matches the input parameter
     * 
     * @param programWorkflowStateId the primary key Int id to match
     * @return a [ProgramWorkflowState] whose identifier matches the passed
     *         [programWorkflowStateId]
     */
    open fun getState(programWorkflowStateId: Int?): ProgramWorkflowState? {
        return states.firstOrNull { it.programWorkflowStateId == programWorkflowStateId }
    }
    
    /**
     * Returns a [ProgramWorkflowState] whose Concept matches the passed concept
     * 
     * @param concept the Concept to match
     * @return Returns a [ProgramWorkflowState] whose [Concept] matches the passed
     *         [concept]
     */
    open fun getState(concept: Concept?): ProgramWorkflowState? {
        return states.firstOrNull { it.concept == concept }
    }
    
    /**
     * Returns a [ProgramWorkflowState] whose Concept name matches the passed name in any
     * Locale
     * 
     * @param name the Concept name to match in any Locale
     * @return a [ProgramWorkflowState] whose [Concept] name matches the passed
     *         [name]
     */
    open fun getState(name: String?): ProgramWorkflowState? {
        return states.firstOrNull { it.concept?.isNamed(name) == true }
    }
    
    /**
     * Returns a [ProgramWorkflowState] whose [Concept] has any [ConceptName] that
     * matches the given [name]
     * 
     * @param name the [ProgramWorkflowState] name, in any Locale
     * @return a [ProgramWorkflowState] which has the passed [name] in any
     *         Locale
     */
    open fun getStateByName(name: String?): ProgramWorkflowState? {
        return states.firstOrNull { it.concept?.isNamed(name) == true }
    }
    
    /**
     * Returns a Set<[ProgramWorkflowState]> including all non-retired ProgramWorkflowStates
     * and all retired ProgramWorkflowStates in this ProgramWorkflow if [includeRetired]
     * is true
     * 
     * @param includeRetired - if false, returns only non-retired [ProgramWorkflowState]
     *            objects in this ProgramWorkflow
     * @return Set<ProgramWorkflowState> - all ProgramWorkflowStates matching input parameters
     */
    open fun getStates(includeRetired: Boolean): Set<ProgramWorkflowState> {
        return if (includeRetired) {
            states.toSet()
        } else {
            states.filterNot { it.retired }.toSet()
        }
    }
    
    /**
     * Returns a Set<[ProgramWorkflowState]> including all ProgramWorkflowStates, sorted by
     * [ConceptName]
     * 
     * @return Set<ProgramWorkflowState> - all ProgramWorkflowStates, sorted by [ConceptName]
     * Should sort names containing numbers intelligently
     */
    open val sortedStates: Set<ProgramWorkflowState>
        get() {
            val naturalComparator = NaturalStrings.getNaturalComparator()
            val stateComparator = compareBy<ProgramWorkflowState> { state ->
                state.concept?.name?.name
            }.let { comparator ->
                Comparator { o1, o2 -> 
                    naturalComparator.compare(
                        o1.concept?.name?.name, 
                        o2.concept?.name?.name
                    )
                }
            }
            return states.sortedWith(stateComparator).toSet()
        }
    
    /**
     * Returns a List<[ProgramWorkflowState]> including all possible next
     * ProgramWorkflowStates, for the passed [PatientProgram] ordered by [ConceptName]
     * 
     * @param patientProgram - The PatientProgram to check
     * @return List<ProgramWorkflowState> - all possible next ProgramWorkflowStates, for the passed
     *         [PatientProgram] ordered by [ConceptName]
     */
    open fun getPossibleNextStates(patientProgram: PatientProgram): List<ProgramWorkflowState> {
        val currentState = patientProgram.getCurrentState(this)
        return sortedStates.filter { state ->
            isLegalTransition(currentState?.state, state)
        }
    }
    
    /**
     * Check whether it is allowable to transition from [fromState] to
     * [toState].
     * 
     * @param fromState [ProgramWorkflowState] to check transition from
     * @param toState [ProgramWorkflowState] to check transition to
     * @return boolean true if it is allowable to transition from [fromState] to
     *         [toState]
     */
    open fun isLegalTransition(fromState: ProgramWorkflowState?, toState: ProgramWorkflowState): Boolean {
        // If there's no current state then we need to move into an initial state
        if (fromState == null) {
            return toState.initial == true
        }
        
        // Does not allow patient to move into the same state
        return fromState != toState
    }
    
    /** @see Object.toString */
    override fun toString(): String {
        return "ProgramWorkflow(id=$programWorkflowId)"
    }
    
    // ******************
    // Property Access
    // ******************
    
    /**
     * @since 1.5
     * @see org.openmrs.OpenmrsObject.getId
     */
    override var id: Int?
        get() = programWorkflowId
        set(id) {
            programWorkflowId = id
        }
    
    /**
     * Gets the number of states which are not retired
     * 
     * @return the total number of non retired states
     * @since 1.9
     */
    open val nonRetiredStateCount: Int
        get() = getStates(false).size
}
